### PR TITLE
[Ready] Automatic Zombie Advancement aka Real-time mode

### PR DESCRIPTION
**Pull request by lucianposton**
This PR adds a game option to enable the Automatic Zombie Advancement feature, which makes game turns pass automatically (while not inside a menu). This causes zombies and other monsters automatically take turns, adding more pressure and mood on gameplay.

The added option is turn duration, in seconds, which allows users to customize how fast turns will automatically pass.
1. 08a9749 - Adds the Automatic Zombie Advancement feature and user option. 
2. ae2813b - AZA mode would send a safemode warning message on every turn duration. This commit adds a boolean state flag, game::safe_mode_warning_logged, to prevent this.
3. edbdd84 - AZA mode would cause the message TTL feature, when enabled, to prematurely evict messages from the log. This commit adds ACTION_TIMEOUT to the action blacklist introduced by PR #16007.
4. e892c34 - AZA mode would interrupt auto-travel via left-clicking. Auto-travel requires 1 click to set the destination and a 2nd left click to begin the auto-travel. If a timeout event came between those 2 clicks, it would clear the set destination

### CODE REVIEW

**Comment by Coolthulhu** [comment]
Consider those:
* Inventory management
* Aiming
* All sorts of menus

Those often aren't doable in the same time scale as walking and fighting.

**Comment by mugling** [comment]
Crafting, sleeping?

**Comment by lucianposton** [comment]
I'm thinking the menus and multi-turn actions like crafting and sleeping shouldn't be affected. In other words, the game would remain paused within menus, and crafting, auto-travel, sleeping, etc would process at rapid pace, just as they do now. Only while traveling in the defaultmode would turns pass at the user-defined rate.

That's what this pr does, anyway. I'm certainly open to other ideas.

**Comment by kevingranade** [comment]
The problem here is the fix ups to make real time work properly could get
really invasive, and I doubt we'd be willing to make enough of them to make
it work properly.
If I'm wrong and it's simple, we can do it, but I'm not willing to announce
we have the feature if it only half works.

**Comment by Chezzo** [comment]
This is a whole lot of fun. I didn't want to like it and I looked for where it didn't work. I don't play real time games. 

I didn't find it. 

I think the beauty is in the simplicity. It is only on that one screen, you can turn it right off by hitting lots of keys, it just adds a little pressure. 

Even with it set to .75, my normal game play speed was faster, and this mode keeps up with it fine. But if you stop, the zombies get closer, and closer, and closer.

It feels a little like Space Invaders! In a really good way.

Tl;dr: Give it a shot. All it adds is mood, really. But if you are like me, you will love the mood it adds.

File: src/options.cpp [hunk 1]
```diff
@@ -1811,6 +1817,7 @@ bool options_manager::save(bool ingame)
     bIngame = ingame;
     const auto savefile = FILENAMES["options"];
 
+    turn_duration = OPTIONS["TURN_DURATION"]; // cache to global due to heavy usage.
```
Comment (kevingranade) [inline src/options.cpp]: Once every 125ms isn't heavy usage, this doesn't warrant a global.

**Comment by Rivet-the-Zombie** [comment]
I like this idea a lot!

**Comment by lucianposton** [comment]
Renamed option to "Automatic Zombie Advancement" at ChestHole's suggestion. :)

**Comment by devilstars** [comment]
Is this mode enabled in experimental branch?

**Comment by lucianposton** [comment]
@devilstars No. I need to fix some things yet. After I do that, if it's accepted, this pull request will be marked as "merged" (instead of open). Then it'll show up in the next experimental build for that day.

**Comment by macrosblackd** [comment]
If anyone is still confused about what this is actually doing, it might be better to describe it as giving player turns a time limit in which they must perform an action (or go into a menu), after which the game  treats things as if the player pressed the wait-one-turn button.

**Comment by Mecares** [comment]
Are you planning to implement a hot key for activating/deactivating this function? Makes it working like a pause button.

**Comment by kevingranade** [comment]
Every menu pauses, why do you need a dedicated pause button?

**Comment by Rivet-the-Zombie** [comment]
I think the idea is that you'd be able to switch between the current turn-based system and the auto-turn system with a single keypress instead of having to go to the options menu to change it. In that fashion it would effectively serve as a pause button.

**Comment by kevingranade** [comment]
I don't care about adding the method, but I'd be very against making it a
bound key by default, those are in very short supply.

**Comment by lucianposton** [comment]
As is, you can simply look around (default x hotkey iirc) to pause the game, so I don't see a great deal of use for a toggle hotkey.

Let me first wrap up this PR to keep its scope small, and later, if it's still a requested feature, we can add an unbound toggle hotkey.

**Comment by lucianposton** [comment]
atm, this PR is blocked on https://github.com/CleverRaven/Cataclysm-DDA/pull/16007, which provides the means to address AZA's issues with the message TTL feature.

I'll rebase this PR on top of #16007. It makes this PR look 3x it's current size, but that'll go away once #16007 is merged.

**Comment by lucianposton** [comment]
Before and after:
before
out

**Comment by lucianposton** [comment]
With #16007 merged, rebasing this PR onto master to simplify the diff.

File: src/game.cpp [hunk 2]
```diff
@@ -2228,7 +2264,13 @@ input_context game::get_player_input(std::string &action)
         }
         inp_mngr.set_timeout(-1);
     } else {
-        while (handle_mouseview(ctxt, action)) {};
+        inp_mngr.set_timeout(125);
+        while (handle_mouseview(ctxt, action)) {
+            if (action == "TIMEOUT" && has_user_turn_timeout_elapsed()) {
```
Comment (kevingranade) [inline src/game.cpp]: Why did you check for turn timeout again here? It should only be set to timeout if a previous check already returned true.

File: src/game.cpp [hunk 3]
```diff
@@ -175,6 +176,34 @@ bool is_valid_in_w_terrain(int x, int y)
     return x >= 0 && x < TERRAIN_WINDOW_WIDTH && y >= 0 && y < TERRAIN_WINDOW_HEIGHT;
 }
 
+namespace {
+
+using namespace std::chrono;
+
+time_point<steady_clock> user_turn_start;
```
Comment (kevingranade) [inline src/game.cpp]: Since this is only active inside the player input method, the variable should live there, I'm looking at merging, so I'll make a follow-up PR that does this.

File: src/game.cpp (hunk 2 above)
Comment (lucianposton) [inline src/game.cpp]: handle_mouseview() returns true => action == "TIMEOUT" || ctxt.get_raw_input().get_first_input() == ERR. Simply going from that post-condition, it's possible that action != "TIMEOUT". It probably wouldn't make an observable difference either way.

I vaguely recall error events when mousing or clicking along the edge of the screen. I'd have to double check that though. That might have been for something unrelated to this code.

File: src/game.cpp (hunk 3 above)
Comment (lucianposton) [inline src/game.cpp]: thanks
